src/hooks/useLibp2p.ts

```diff
@@ -0,0 +1,157 @@
+import { Connection } from '@libp2p/interface-connection';
+import { PeerInfo } from '@libp2p/interface-peer-info';
+import { StreamHandler } from '@libp2p/interface-registrar';
+import { ethers } from 'ethers';
+import { pipe } from 'it-pipe';
+import { nodeConfig } from 'lib/utils/node_config';
+import { createLibp2p } from 'libp2p';
+import { useCallback, useEffect } from 'react';
+import { setLibp2p } from 'store/app/actions';
+import { setArchaeologistConnection, setArchaeologistOnlineStatus } from 'store/embalm/actions';
+import { useDispatch, useSelector } from 'store/index';
+
+export function useLibp2p() {
+  const dispatch = useDispatch();
+  const libp2pNode = useSelector(s => s.appState.libp2pNode);
+  const archaeologists = useSelector(s => s.embalmState.archaeologists);
+
+  const onPeerDiscovery = useCallback(
+    (evt: CustomEvent<PeerInfo>) => {
+      const peerId = evt.detail.id.toString();
+      dispatch(setArchaeologistOnlineStatus(peerId, true));
+    },
+    [dispatch]
+  );
+
+  const onPeerConnect = useCallback(
+    (evt: CustomEvent<Connection>) => {
+      const peerId = evt.detail.remotePeer.toString();
+      dispatch(setArchaeologistConnection(peerId, evt.detail));
+    },
+    [dispatch]
+  );
+
+  const onPeerDisconnect = useCallback(
+    (evt: CustomEvent<Connection>) => {
+      const peerId = evt.detail.remotePeer.toString();
+      dispatch(setArchaeologistOnlineStatus(peerId, false));
+      dispatch(setArchaeologistConnection(peerId, null));
+    },
+    [dispatch]
+  );
+
+  const handleStream: StreamHandler = useCallback(
+    ({ stream }) => {
+      pipe(stream, async function (source) {
+        for await (const msg of source) {
+          const decoded = new TextDecoder().decode(msg);
+          console.info(`received message ${decoded}`);
+
+          const archConfigJson: Record<string, any> = JSON.parse(decoded);
+
+          const signerAddress = ethers.utils.verifyMessage(
+            JSON.stringify({
+              encryptionPublicKey: archConfigJson.encryptionPublicKey,
+              peerId: archConfigJson.peerId,
+            }),
+            archConfigJson.signature
+          );
+
+          const i = archaeologists.findIndex(a => a.profile.archAddress === signerAddress);
+
+          if (i !== -1) {
+            const arch = archaeologists[i];
+            // TODO: Determine if there's a better way to be certain of origin's peerId
+            if (arch.profile.peerId.toString() !== archConfigJson.peerId) {
+              // This is POSSIBLE, but in practice shouldn't ever happen.
+              // But that's how you know it'll DEFINITELY happen eh? Sigh.
+              console.error('Peer ID mismatch'); // TODO: Handle this problem better. Relay feedback to user.
+              return;
+            }
+
+            console.log('got public key of', arch.profile.peerId);
+
+            arch.publicKey = archConfigJson.encryptionPublicKey;
```

Comment:
I didn't actually run the code to verify this, but doesn't this also need a `updateArchProperty` sort of call to actually update the `publicKey` ... or perhaps I'm still not 100% sure on how the store works ...

Reply:
Ah good catch - there should be an update here, unless that's being handled somewhere else? @knoll3